File: src/main/java/com/github/tianchenghang/agent/ManusAgent.java
package com.github.tianchenghang.agent;

import com.github.tianchenghang.advisor.LoggerAdvisor;
import org.springframework.ai.chat.client.ChatClient;
import org.springframework.ai.chat.model.ChatModel;
import org.springframework.ai.tool.ToolCallback;
import org.springframework.stereotype.Component;

@Component
public class ManusAgent extends ToolCallAgent {

    public ManusAgent(ToolCallback[] allTools, ChatModel dashscopeChatModel) {
        super(allTools);
        this.setName("Code Master");
        var SYSTEM_PROMPT
                = """
              You are Code Master, an all-capable AI assistant, aimed at solving any task presented by the user.
              You have various tools at your disposal that you can call upon to efficiently complete complex requests.
              """;
        this.setSystemPrompt(SYSTEM_PROMPT);
        var NEXT_STEP_PROMPT
                = """
              Based on user needs, proactively select the most appropriate tool or combination of tools.
              For complex tasks, you can break down the problem and use different tools step by step to solve it.
              After using each tool, clearly explain the execution results and suggest the next steps.
              If you want to stop the interaction at any point, use the `doTerminate` tool/function call.
              """;
        this.setNextStepPrompt(NEXT_STEP_PROMPT);
        this.setMaxSteps(20);
        var chatClient
                = ChatClient.builder(dashscopeChatModel).defaultAdvisors(new LoggerAdvisor()).build();
        this.setChatClient(chatClient);
    }
}

File: server/src/agent/manus-agent.ts
import { StructuredTool } from '@langchain/core/tools';
import { ChatModel } from '../llm/chat-model';
import { ToolExecutor } from '../tools/types';
import { ToolCallAgent } from './tool-call-agent';

const SYSTEM_PROMPT = `
  You are an all-capable AI assistant, aimed at solving any task presented by the user.
  You have various tools at your disposal that you can call upon to efficiently complete complex requests.
  Use chinese.
`;

const NEXT_STEP_PROMPT = `
  Based on user needs, proactively select the most appropriate tool or combination of tools.
  For complex tasks, you can break down the problem and use different tools step by step to solve it.
  After using each tool, clearly explain the execution results and suggest the next steps.
  If you want to stop the interaction at any point, use the \`TerminateTool\` tool/function call.
  Use chinese.
`;

export class ManusAgent extends ToolCallAgent {
  constructor(
    tools: StructuredTool[],
    toolExecutor: ToolExecutor,
    chatModel: ChatModel,
    ragContext?: string,
  ) {
    super(tools, toolExecutor);
    this.name = ManusAgent.name;
    this.systemPrompt = ragContext
      ? `${SYSTEM_PROMPT}\n\n${ragContext}`
      : SYSTEM_PROMPT;
    this.nextStepPrompt = NEXT_STEP_PROMPT;
    this.maxSteps = 20;
    this.chatModel = chatModel;
  }
}
